feat: pallet benchmarking interactive UI

chungquantin opened the pull request:
# Description

The PR is a part of a bigger PR #424. It implements the full flow of the `pop bench pallet` including new key features: 
- Interactive UI to manage parameters before running the benchmark. 
- Sourcing the `frame-omni-bencher` binary #428. This is required to list pallets and extrinsics. [Explained here](https://github.com/r0gue-io/pop-cli/pull/420#discussion_r1974726796)
- Refactor to use our own version of `PalletCmd` (called `BenchmarkPallet`) instead of `frame-benchmarking-cli::PalletCmd` due to required parameters issue. [Explained here](https://github.com/r0gue-io/pop-cli/pull/420#discussion_r1974742591)

## How to test the PR? 

```
pop bench pallet
```

If arguments are provided like 

```
pop bench pallet --runtime=path-to-runtime --pallet= --extrinsic=
```

We skip the runtime, the pallet and the extrinsic.

To skip the parameter menu, user can specify flag `--skip` then the parameter menu can be skipped and go directly to run the bencmarking. 

```
pop bench pallet --skip
```

Test single line command
```
pop bench pallet --pallet=pallet_balances --extrinsic=burn_allow_death,burn_keep_alive --steps=50 --repeat=20 --external-repeat=1 --db-cache=1024 --map-size=1000000 --additional-trie-layers=2 --output-pov-analysis=median-slopes --runtime=/Users/chungquantin/Developer/pop-network/pop-cli/tests/runtimes/base_parachain_benchmark.wasm --genesis-builder=runtime --genesis-builder-preset=development
```

## Help command

<img width="783" alt="Screenshot 2025-03-05 at 11 56 16" src="https://github.com/user-attachments/assets/35b04dc8-bd76-4391-846d-89a35581d4b3" />

### Full flow breakdown

Note: If `--list` or `--json-output` is provided, we run the command by `frame-benchmarking-cli` directly instead of go through the implementation in pop-cli. 

#### Stage 1: Locate runtime or input runtime path manually
- 1.1. Run the command in a parachain project, which contains `runtime` or `runtimes` (Hyperbridge) folder. This falls to the first stage [**_"Stage 1: Locating and building a runtime binary | Scenario 1"_** of the previous PR](https://github.com/r0gue-io/pop-cli/pull/411). 

<img width="100%" alt="Screenshot 2025-02-28 at 17 05 25" src="https://github.com/user-attachments/assets/fdeae8e7-05cc-4c29-8b6b-24aede3fe69a" />

- 1.2. Run the command anywhere (no a parachain project), which does contains `runtime` or `runtimes` (Hyperbridge) folder. This falls to the first stage [**_"Stage 1: Locating and building a runtime binary | Scenario 2"_** of the previous PR](https://github.com/r0gue-io/pop-cli/pull/411). 

<img width="100%" alt="Screenshot 2025-02-28 at 17 03 33" src="https://github.com/user-attachments/assets/fa04de1c-c962-4f4d-98fa-f01fc1ba1d50" />

- 1.3. Inputting the runtime manually. User can input the path to parachain project or path to WASM binary file directly.

NOTE: After the path is located and there is no WASM file found in the `target/<profile>/wbuild/<runtime-name>.wasm`, we will build the runtime with `runtime-benchmarks` feature first.

### Stage 2: Select genesis builder policy & preset

See full breakdown in #411 

### Stage 3: Select pallets (Single Select) & extrinsics (Multiple Select)

At this stage, we will check if the binary of `frame-omni-bencher` exists on the user's machine. [Explained here](https://github.com/r0gue-io/pop-cli/pull/420#discussion_r1974726796)

Note: We only call `frame-omni-bencher` to load pallets and extrinsics once a session. 

- 3.1. **_"🔎 Search for pallets by name"_**: User can specify the pallet name to search for the pallets (fuzzy searching).

For the extrinsic selection, we only allow user to select if there are pallet specified. If all pallet are selected (`--pallet=*`), we skip the next step. 

- 3.2. **_"🔎 Search for extrinsics by name"_**: User can specify the extrinsic name to search for the extrinsics (fuzzy searching). 

<div style="display:flex;">
<img width="43%" alt="Screenshot 2025-03-05 at 11 50 10" src="https://github.com/user-attachments/assets/360f196c-3676-4743-aad5-90d54e642445" />
<img width="43%" alt="Screenshot 2025-03-05 at 11 51 05" src="https://github.com/user-attachments/assets/b0a116e1-65b2-4843-8972-52ea693bd2de" />
</div>

### Stage 4: Parameter menu

User can update the parameters by selecting an option in the parameter menu. Some arguments will be disabled depend on the status of other parameters. See `is_disabled` method. For example: 
- we don't allow selecting extrinsics if all pallets are selected.
- we don't allow update the genesis builder preset if policy is equal to `GenesisBuilderPolicy::None`

<img width="100%" alt="Screenshot 2025-02-28 at 17 15 29" src="https://github.com/user-attachments/assets/09e76c67-8645-4c5a-b9da-60e4ead0ba56" />

### Stage 5: Input the output path

This stage is optional, if no value is provided, we don't write to any file. 

<img width="100%" alt="Screenshot 2025-02-28 at 17 19 30" src="https://github.com/user-attachments/assets/a3dd7927-7505-46c2-a38d-05871c4bcf8d" />

### Stage 6: Print the single line command 

<img width="100%" alt="Screenshot 2025-02-28 at 17 20 08" src="https://github.com/user-attachments/assets/1ddceee2-7f0a-4c5a-b733-7e89bc3f206e" />



[sc-2882]

chungquantin commented on crates/pop-parachains/src/bench.rs:
```diff
@@ -33,7 +53,143 @@ pub fn get_runtime_path(parent: &Path) -> anyhow::Result<PathBuf> {
+///
+/// # Arguments
+/// * `runtime_path` - Path to the runtime WASM binary.
+pub fn load_pallet_extrinsics(runtime_path: &Path) -> anyhow::Result<PalletExtrinsicsRegistry> {
```
- `Approach 1 (Subwasm + FRAME metadata)`: Current approach of this PR: to retrieve the metadata from Substrate runtime binary is to use `ac-node-api` of `substrate-api-client` and `subwasm`'s WASM loader crate to load the pallets. However, there are issues that the pallet name is represented in the defined name on the runtime level, not the crate name (e.g. `Timestamp` instead of `pallet_timestamp`), hence, when pass it to the `--pallet=` argument, the `frame-benchmarking-cli` can't identify the pallet to benchmark. On the other hand, `pallet::hooks` are not included in the Substrate runtime metadata. But hooks are also benchmarked, so we need to find a way to get them. 

- `Approach 2 (List pallets using frame-benchmarking-cli)`: See branch `chungquantin/feat-pallet_overwrite_stdio`. This approach works perfectly and it was the original approach. This approach is we need to overwrite the stdout. So that when `println!` is called in `frame-benchmarking-cli`, the output is captured and returned back for processing. We use external crate https://github.com/elichai/stdio-override to redirect the stdout. However, the issue happens on the CI when there's a conflict with the stdout. Code to the issue: https://github.com/elichai/stdio-override/blob/master/src/lib.rs#L151

```
thread 'bench::tests::search_extrinsics_works' panicked at /home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/stdio-override-0.2.0/src/lib.rs:236:13:
Stdout override reset out of order!
```

chungquantin replied:
Update: Follow the second approach but no longer overwrite the `stdout`, the main issue is with the CI. CI environment spawns their own child processes and overwrite the stdout with their own buffer. Hence, it causes the implementation fails. On the other hand, overwrite stdout might not be a stable approach for cross platforms. 

Solution to that is to use the `frame-omni-bencher` and source it from the remote repository, implementation in #428

chungquantin commented on crates/pop-parachains/src/bench.rs:
```diff
@@ -33,7 +53,143 @@ pub fn get_runtime_path(parent: &Path) -> anyhow::Result<PathBuf> {
+///
+/// # Arguments
+/// * `cmd` - Command to benchmarking extrinsic weights of FRAME pallets.
+pub fn print_pallet_command(cmd: &PalletCmd) -> String {
```
There are arguments which are not covered yet:
- `wasm_method` due to private type `WasmExecutionMethod`
- `wasm_instantiation_strategy` due to private type `WasmtimeInstantiationStrategy` 
- `default_pov_mode` due to private type `PovEstimationMode`
- `hostinfo_params` due to private type `HostInfoParams`

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -51,78 +80,513 @@ impl BenchmarkPallet {
 }
 
+#[derive(Clone, Copy, EnumIter, EnumMessageDerive, Eq, PartialEq)]
+enum BenchmarkPalletMenuOption {
```
Not all arguments are listed in the menu. 
- `wasm_method` due to private type `WasmExecutionMethod`
- `wasm_instantiation_strategy` due to private type `WasmtimeInstantiationStrategy` 
- `default_pov_mode` due to private type `PovEstimationMode`
- `hostinfo_params` due to private type `HostInfoParams`

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -51,78 +74,513 @@ impl BenchmarkPallet {
+		let spinner = spinner();
+		spinner.start("Fetching pallets and extrinsics from your runtime...");
+		let runtime_path = get_runtime_argument(cmd)?;
+		log::set_max_level(LevelFilter::Off);
```
No longer needed

chungquantin commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,352 @@
+) -> anyhow::Result<PalletExtrinsicsRegistry> {
+	let temp_file = NamedTempFile::new()?;
+
+	let mut process = run_benchmarking_with_binary(
```
The reason why we need to download the binary.

chungquantin commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,352 @@
+
+	// Process the captured output and return the pallet extrinsics registry.
+	let mut registry = PalletExtrinsicsRegistry::new();
+	let lines: Vec<String> = output.split("\n").map(String::from).skip(1).collect();
```
`.skip(1)` because first line is `pallet, extrinsic`

chungquantin commented on crates/pop-parachains/src/bench.rs:
Move all the code to `bench/mod.rs` to keep it more organized with separated logic.

chungquantin commented on crates/pop-parachains/Cargo.toml:
```diff
@@ -37,6 +37,7 @@ zombienet-sdk.workspace = true
 # Benchmarking
 cumulus-primitives-proof-size-hostfunction.workspace = true
 frame-benchmarking-cli.workspace = true
+fuzzy-matcher = "0.3.7"
```
Add this crate to fuzzy searching the pallets and extrinsics by name

chungquantin replied:
Removed this crate in #442

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,71 +1,743 @@
+
+const ALL_SELECTED: &str = "*";
+const MAX_EXTRINSIC_LIMIT: usize = 15;
+const MAX_PALLET_LIMIT: usize = 20;
```
To keep the output on terminal clean, I believe we should limit the number of displayed pallets

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,71 +1,743 @@
+const MAX_PALLET_LIMIT: usize = 20;
+
+#[derive(Args, Default)]
+pub(crate) struct BenchmarkPallet {
```
We implement our own version of `frame-benchmarking-cli::Pallet` because two important fields `--pallet` and `--extrinsic` are required in the original struct. Hence, the parsing check fails when user call `pop bench command` without those arguments. The expected behaviour is to allow user to select pallets and extrinsics if the arguments are not provided. One solution is to use `ignore_errors = true` provided by `clap-rs` but it would cause an issue with the `help` command. Discussed here: https://github.com/clap-rs/clap/issues/5919

Missing below arguments defined in `frame-benchmarking-cli::PalletCmd`:

### Parameters too advanced and rarely used so I don't include them

- `pub hostinfo_params: HostInfoParams`

Parameters to configure how the host info will be determined.

- `pub default_pov_mode: command::PovEstimationMode`

The PoV estimation mode of a benchmark if no `pov_mode` attribute is present.

### Not important to Pop CLI

- `pub no_csv_header: bool`

Don't include csv header when listing benchmarks.

- `pub execution: Option<String>` (Deprecated)

DEPRECATED: This argument has no effect.

- `pub unsafe_overwrite_results: bool`

Allow overwriting a single file with multiple results.
This exists only to restore legacy behaviour. It should never actually be needed.

- `quiet: bool`

Do not print a summary at the end of the run.

- `disable_proof_recording: bool`

Do not enable proof recording during time benchmarking.

### Consider to add

- `pub wasm_method: WasmExecutionMethod`

Method for executing Wasm runtime code.

- `pub wasmtime_instantiation_strategy: WasmtimeInstantiationStrategy`

The WASM instantiation method to use. Only has an effect when `wasm-execution` is set to `compiled`.

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,71 +1,797 @@
+	#[arg(long)]
+	pub extra: bool,
 
-const GENESIS_BUILDER_NO_POLICY: &str = "none";
```
As we implement our own `GenesisBuilderPolicy` enum, this is no longer needed.

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,71 +1,797 @@
+
+	/// If this is set to true, no parameter menu pops up
+	#[arg(long = "skip")]
+	skip_menu: bool,
```
Additional argument. When provided, parameter menu won't pop up.

AlexD10S commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,449 @@
+// SPDX-License-Identifier: GPL-3.0
```
Not requesting the change but curious why in `common` folder? The methods are only used in the bench functionality

chungquantin replied:
These methods will be shared by other bench commands like `overhead` (See #432). So I think it makes more sense to gather them in the common folder.

chungquantin commented on Cargo.toml:
```diff
@@ -76,6 +77,7 @@ git2_credentials = "0.13.0"
 # benchmarking
 cumulus-primitives-proof-size-hostfunction = "0.10.0"
 frame-benchmarking-cli = "46.0.0"
+fuzzy-matcher = "0.3.7"
```
Seems like the implementation in cliclack looks better. Created a PR here to add the feature to `Cli`: https://github.com/r0gue-io/pop-cli/pull/435

chungquantin replied:
Refactored the code to support filter mode in the pallet benchmarking: https://github.com/r0gue-io/pop-cli/pull/442

chungquantin commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,341 @@
+// SPDX-License-Identifier: GPL-3.0
+
+use clap::Parser;
+use frame_benchmarking_cli::PalletCmd;
```
For this PR, agree that we don't need it. But for other commands, we don't want to fork the command struct if possible, so the crate `frame-benchmarking-cli` is still required to implement others. Reference: #432

AlexD10S replied:
It makes sense to keep it then, thanks!

chungquantin commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,341 @@
+	Runtime,
+}
+
+impl ToString for GenesisBuilderPolicy {
```
Resolved by refactoring to `impl Display`

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,334 +1,1251 @@
+
+		// Display the benchmarking command.
+		let mut message = self.display();
+		if self.skip_menu {
```
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/36130e724354c32229ea66379bd7fc3fcdee0080

chungquantin commented on Cargo.toml:
```diff
@@ -37,6 +37,7 @@ tempfile = "3.10"
 thiserror = "1.0.58"
 tokio-test = "0.4.4"
 toml = "0.5.0"
+which = "4.4"
```
Resolved by removing the `which` crate and implement which with `cmd` instead. https://github.com/r0gue-io/pop-cli/pull/420/commits/112f98272233e17d38eca656047e3e8acc53fc60.

I don't find the method `check_command_exists` matches my needs and I found it simpler to just call the `cmd` directly. Note it here that the logic can be refactored to standalone function for future us to get path of `command`.

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,334 +1,1251 @@
-	}
-	Ok(())
+#[derive(Clone, Copy, EnumIter, EnumMessageDerive, Eq, PartialEq)]
+enum BenchmarkPalletMenuOption {
```
Not sure what is the better way to do this. One way is to build a similar experience is having our own macro to construct the menu from struct. But the drawback that it's harder to manage. There are fields in the `BenchmarkPallet` we don't want to display on the menu.

AlexD10S commented on crates/pop-parachains/src/bench/binary.rs:
https://github.com/r0gue-io/polkadot/pull/1

chungquantin commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,341 @@
+use strum_macros::{EnumIter, EnumMessage as EnumMessageDerive};
+use tempfile::NamedTempFile;
+
+pub mod binary;
```
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/b5747fb90b20ac04de1222f3143a8b058f0e034b

chungquantin commented:
> Well done on sourcing the binary! I wonder if we can now remove the `frame-benchmarking` library now and fix the CI issues.
> 
> I've left some comments, but I still need to do a more in-depth review of `common/bench.rs`, especially `bench/pallet.rs`, particularly the tests. Since the coverage CI isn’t working, the tests require extra attention.
> 
> I also need to do more manual testing because I found errors in the searcher:
> 
> ```
> ▲  NOTE: the `pop bench pallet` is not yet battle tested - double check the results.
> │  
> ◇  Select the runtime:
> │  pop-runtime-testnet 
> │
> ◇  🔎 Search for pallets by name ("*" to select all)
> │  balances
> │
> Error: No items added to the list
> ```

Issue with the pallet search is resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/6681706b7e326778144e7cd2c49fe213a3df5233. There was a missing `--genesis-builder=none` when running benchmark with binary.

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,334 +1,1251 @@
+	// Prompt for pallet search input.
+	let input = cli
+		.input(r#"🔎 Search for pallets by name ("*" to select all)"#)
+		.placeholder("balances")
```
https://github.com/r0gue-io/pop-cli/pull/420/commits/97af1173e49fbb2659f24272e4f8ec2eb314e337 removes the ("*" to select all) logic and confirm with the user first before listing pallets

<img width="572" alt="Screenshot 2025-03-05 at 10 52 32" src="https://github.com/user-attachments/assets/fadf094b-0d84-421b-bf91-64b62e44c886" />

chungquantin commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,449 @@
+///
+/// # Arguments
+/// * `cli`: Command line interface.
+pub fn guide_user_to_input_runtime_path(cli: &mut impl Cli) -> anyhow::Result<PathBuf> {
```
Agree. This method is redundant. Removed.

chungquantin commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,238 @@
+		Some(temp_file.as_file()),
+		vec![
+			&format!("--runtime={}", runtime_path.display()),
+			"--genesis-builder=none",
```
We need this flag otherwise error thrown in the child process.

chungquantin commented on crates/pop-common/src/helpers.rs:
```diff
@@ -55,6 +55,20 @@ pub fn prefix_with_current_dir_if_needed(path: PathBuf) -> PathBuf {
+/// # Arguments
+/// * `base` - The base directory to compare against.
+/// * `full` - The full path to be shortened.
+pub fn get_relative_or_absolute_path(base: &Path, full: &Path) -> PathBuf {
```
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/7641a8a3fa76ba6cc40a16dff0652a49ec85ed66

chungquantin commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,449 @@
+		.default_input("./runtime")
+		.placeholder("./runtime")
+		.interact()
+		.map(PathBuf::from)
```
Resolved!

chungquantin commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,449 @@
+		.input("Please provide the path to the runtime or parachain project.")
+		.required(true)
+		.default_input("./runtime")
+		.placeholder("./runtime")
```
Resolved!

Daanvdplas commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+///
+/// # Arguments
+/// * `cli`: Command line interface.
+/// * `target_path`: A reference to the target path.
```
```suggestion
/// * `target_path`: The target path.
```

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

Daanvdplas commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+				target_path.display()
+			))?;
+			let input: PathBuf = cli
+				.input("Please provide the path to the runtime or parachain project.")
```
```suggestion
				.input("Please provide the path to the runtime.")
```

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

Daanvdplas commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+		},
+	};
+
+	// If there is no TOML file exist, list all directories in the "runtime" folder and prompt the
```
```suggestion
	// If a TOML file does not exist, list all directories in the "runtime" folder and prompt the
```

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

Daanvdplas commented on crates/pop-cli/src/common/mod.rs:
```diff
@@ -1,5 +1,7 @@
 // SPDX-License-Identifier: GPL-3.0
 
+#[cfg(feature = "parachain")]
+pub mod bench;
```
Public docs

chungquantin replied:
Added comment. But other public modules in the file is not commented as well.

Daanvdplas commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,238 @@
+	cumulus_primitives_proof_size_hostfunction::storage_proof_size::HostFunctions,
+);
+
+/// Type alias for records where the key is the pallet name and the value is a array of its
```
```suggestion
/// Type alias for records where the key is the pallet name and the value is an array of its
```

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

Daanvdplas commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,238 @@
+/// Get genesis builder preset names of the runtime.
+///
+/// # Arguments
+/// * `binary_path` - Path to the runtime WASM binary.
```
Suggesting this removal so that when the time comes where runtimes will be ran on PVM we don't have to make this change:
```suggestion
/// * `binary_path` - Path to the runtime binary.
```

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

Daanvdplas commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,238 @@
+	genesis_config_builder.preset_names().map_err(|e| anyhow::anyhow!(e))
+}
+
+/// Get the runtime folder path and throws error if not exist.
```
```suggestion
/// Get the runtime folder path and throws error if it does not exist.
```

Daanvdplas commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,238 @@
+		.ok_or_else(|| anyhow::anyhow!("No runtime found"))
+}
+
+/// Runs FRAME pallet benchmarks using `frame-benchmarking-cli`.
```
```suggestion
/// Runs pallet benchmarks using `frame-benchmarking-cli`.
```

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

Daanvdplas commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,238 @@
+		.map_err(|e| anyhow::anyhow!("Failed to run benchmarking: {}", e))
+}
+
+/// Loads a mapping of pallets and their associated extrinsics from the runtime WASM binary.
```
```suggestion
/// Loads a mapping of pallets and their associated extrinsics from the runtime binary.
```

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

Daanvdplas commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,238 @@
+/// Loads a mapping of pallets and their associated extrinsics from the runtime WASM binary.
+///
+/// # Arguments
+/// * `runtime_path` - Path to the runtime WASM binary.
```
```suggestion
/// * `runtime_path` - Path to the runtime binary.
```

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

Daanvdplas commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,238 @@
+		Ok(())
+	}
+
+	fn get_mock_runtime_path(with_runtime_benchmarks: bool) -> PathBuf {
```
Is this double with: https://github.com/r0gue-io/pop-cli/blob/5181c5a64b2e0933b1c406b3499babc8d2db1fc9/crates/pop-cli/src/common/bench.rs#L283

chungquantin replied:
Yes they are the same but I don't find trying to make this shared across packages seems reasonable. It is only a method to get mock runtime and the path can be different between folder levels.

Daanvdplas reviewed (commented):
Thanks for the super clear information provided! I was able to immediately start reviewing without any hurdles, fantastic!!!

Also well done on the PR and research that you have put into it!

I have a question and some commands that were failing for me, I have gone through all the code but haven't been able to test the functionality yet so that will come next time. 

Why `pop bench pallet` and not `pop bench`?

Some commands that were failing:
- `pop bench pallet --list` not so clear error:
<img width="1178" alt="Screenshot 2025-03-05 at 15 36 21" src="https://github.com/user-attachments/assets/a60a9377-69fd-485a-a566-b389dd724d3d" />

Resolved by not throwing error anymore. Listing pallets for non `runtime-benchmarks` feature runtime binary works normally. 

- `pop bench pallet`:
<img width="957" alt="Screenshot 2025-03-05 at 16 30 59" src="https://github.com/user-attachments/assets/f4ed7540-2cd7-4ea4-a0d6-4f94ca75f19d" />

@chungquantin Resolved: 
<img width="815" alt="Screenshot 2025-03-06 at 19 02 56" src="https://github.com/user-attachments/assets/0abd89e7-4d60-42ad-b8ca-af48e2c73b5e" />

- `pop bench pallet --runtime=runtime --pallet= --extrinsic=` in `pop-node` folder:
<img width="990" alt="Screenshot 2025-03-05 at 16 40 23" src="https://github.com/user-attachments/assets/56d12ddc-4ecb-4b96-a6ac-e3154d2dc7f3" />


@chungquantin Resolved: 
<img width="568" alt="Screenshot 2025-03-06 at 19 07 22" src="https://github.com/user-attachments/assets/63de62c9-180f-4c5c-963f-fb36a0673930" />

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,380 +1,1342 @@
+		if self.all {
+			self.pallet = Some(ALL_SELECTED.to_string());
+			self.extrinsic = Some(ALL_SELECTED.to_string());
+			self.all = false;
```
The issue is if you don't set it to false, `--all` will override any value of `--pallet` and `--extrinsic`

AlexD10S commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+	Ok(match cmd("which", &["frame-omni-bencher"]).stdout_capture().run() {
+		Ok(output) => {
+			let path = String::from_utf8(output.stdout)?;
+			PathBuf::from(path.replace("\n", ""))
```
Not requesting a change, but I belieive you can use `path.trim()` here for the same

chungquantin replied:
Agree. Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/89791b8657a36ac42a0b9adee27fb0b5a3225e75

AlexD10S commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+	cli: &mut impl Cli,
+	project_path: &PathBuf,
+) -> anyhow::Result<PathBuf> {
+	let runtimes = fs::read_dir(project_path).expect("No project found");
```
```suggestion
	let runtimes = fs::read_dir(project_path)?;
```

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/89791b8657a36ac42a0b9adee27fb0b5a3225e75

AlexD10S commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,238 @@
+/// # Arguments
+/// * `binary_path` - Path to the runtime WASM binary.
+pub fn get_preset_names(binary_path: &PathBuf) -> anyhow::Result<Vec<String>> {
+	let binary = fs::read(binary_path).expect("No runtime binary found");
```
```suggestion
	let binary = fs::read(binary_path)?;
```

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/89791b8657a36ac42a0b9adee27fb0b5a3225e75

AlexD10S commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,238 @@
+		match s {
+			s if s == *"none" => GenesisBuilderPolicy::None,
+			s if s == *"runtime" => GenesisBuilderPolicy::Runtime,
+			_ => unreachable!(),
```
`unreachable! `panics right? Better to throw an error

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/89791b8657a36ac42a0b9adee27fb0b5a3225e75

AlexD10S commented on crates/pop-parachains/src/bench/mod.rs:
```diff
@@ -0,0 +1,341 @@
+/// # Arguments
+/// * `args` - Arguments to pass to the benchmarking command.
+pub fn generate_benchmarks(args: Vec<String>) -> anyhow::Result<()> {
+	let cmd = PalletCmd::try_parse_from([vec!["".to_string()], args].concat())
```
Thanks for the great explanation. Your approach makes more sense

AlexD10S commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,380 +1,1342 @@
+) -> anyhow::Result<String> {
+	let pallets = pallets(registry, excluded_pallets);
+	if pallets.is_empty() {
+		return Err(anyhow::anyhow!("No pallets found for the runtime"));
```
I noticed that when I run `pop bench pallet` in a parachain built without `--features runtime-benchmarks`, I get the error:
`"Error: No pallets found for the runtime."`
It might be better to improve this error message by suggesting that the parachain should be built with -`-features runtime-benchmarks`. What do you think?

chungquantin replied:
I agree. This could be improved. Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/ff2f39d1f59d7bb7766d604804d2c18fc19bfcc3

Now if the runtime-benchmarks feature is not enabled, it will throw error: 

<img width="813" alt="Screenshot 2025-03-06 at 18 57 24" src="https://github.com/user-attachments/assets/fb54348c-0f36-441d-94e5-fa09aa844cf5" />

AlexD10S replied:
Really cool!

chungquantin commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+///
+/// # Arguments
+/// * `cli`: Command line interface.
+/// * `cache_path`: The cache directory path.
```
<img width="563" alt="Screenshot 2025-03-06 at 13 44 02" src="https://github.com/user-attachments/assets/ad063453-d9d0-48a5-b689-75ed78a6d4e0" />


To keep it consistent with other functions.

chungquantin replied:
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df by removing the argument

chungquantin commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+///
+/// # Arguments
+/// * `cli`: Command line interface.
+/// * `mode`: A reference to the profile mode.
```
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

chungquantin commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+///
+/// # Arguments
+/// * `cli`: Command line interface.
+/// * `runtime_path`: A reference to the runtime path.
```
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

chungquantin commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+	let target_path = mode.target_directory(&cwd).join("wbuild");
+	let runtime_path = guide_user_to_select_runtime_path(cli, &cwd)?;
+
+	// Return immediately if the user has specified a path to the runtime binary.
```
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

chungquantin commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+}
+
+/// Ensure that the runtime WASM binary exists and is ready for use. If the binary is not found, it
+/// triggers a build process.
```
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

chungquantin commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+
+/// Checks the status of the `frame-omni-bencher` binary, use the local binary if available.
+/// Otherwise, sources it if necessary, and prompts the user to update it if the existing binary in
+/// cache is not the latest version.
```
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

chungquantin commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,380 +1,1342 @@
+
+#[derive(Args)]
+pub(crate) struct BenchmarkPallet {
+	/// Select a FRAME Pallet to benchmark, or `*` for all (in which case `extrinsic` must be `*`).
```
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/40a80c6c0bcd771ef886c3e5e84b67cc6e9634df

chungquantin commented on crates/pop-cli/src/common/bench.rs:
```diff
@@ -0,0 +1,456 @@
+/// # Arguments
+/// * `cli`: Command line interface.
+/// * `mode`: A reference to the profile mode.
+pub fn ensure_runtime_binary_exists(cli: &mut impl Cli, mode: &Profile) -> anyhow::Result<PathBuf> {
```
Resolved in https://github.com/r0gue-io/pop-cli/pull/420/commits/8e3fee344e1d7beb1c80fdf51913c5b781a8ac31

AlexD10S commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,380 +1,1342 @@
+				match option.update_arguments(self, &mut registry, cli).await {
+					Ok(true) => break,
+					Ok(false) => continue,
+					Err(e) => cli.info(e)?,
```
I don't mind to throw an error, it was more about the message` cliclack::log::error` or maybe a `cliclack::log::warning` instead of an `info`.

chungquantin replied:
Ah I see, apologies I did not aware of the feature in `cliclack`

chungquantin replied:
Resolved by changing it to `cliclack::log::error`

Daanvdplas commented on crates/pop-cli/src/commands/bench/pallet.rs:
```diff
@@ -1,380 +1,1342 @@
+
+	/// Comma separated list of pallets that should be excluded from the benchmark.
+	#[arg(long, value_parser, num_args = 1.., value_delimiter = ',')]
+	pub exclude_pallets: Vec<String>,
```
And with the flag `all`? If that is not allowed we could add the `conflicts_with`

Daanvdplas commented:
- I did `pop bench pallet` and gave input to points 0 to 4, now in the parameter menu they are asked again:
<img width="949" alt="Screenshot 2025-03-07 at 16 19 28" src="https://github.com/user-attachments/assets/bcc68b3f-368a-4d6c-942a-ad0a20b5abf5" />
If it is an easy fix, can we eliminate the things that are provided by the user from the menu?

- Could we ask if the user would like to go through the parameter menu? The skip flag is already a great flag!
- Loading the menu is also slow, do we know why?
- Can we remove the word FRAME here?
<img width="729" alt="Screenshot 2025-03-07 at 16 25 09" src="https://github.com/user-attachments/assets/4d252cf4-88bb-49b1-8f01-e980b3bb902c" />

Could we add a newline before outputting the generated command (great feature!):
<img width="1436" alt="Screenshot 2025-03-07 at 16 39 25" src="https://github.com/user-attachments/assets/1c6f01dd-a5d0-4516-ad67-94a7e316881a" />

When we can choose the extrinsics to benchmark, can we add `(select with space)`?
<img width="483" alt="Screenshot 2025-03-07 at 16 42 29" src="https://github.com/user-attachments/assets/6fc0c7a3-c9ab-490e-af13-f5c2acab13ee" />

Daanvdplas reviewed (approved):
I left some UX improvements / suggestions, but wow what an amazing PR!

chungquantin commented:
@Daanvdplas 

> If it is an easy fix, can we eliminate the things that are provided by the user from the menu?

This goes against the benefit of the parameter menu that we defined in the spec. The benefit of the menu is to allow user to update the parameter again if they mistype. Reason for that because benchmarking has more parameters compared to other commands. Hence, this gives the user more control to the parameters of the command. 

> Could we ask if the user would like to go through the parameter menu? The skip flag is already a great flag!

Agree with this suggestion. Added in the next commit. 

> When we can choose the extrinsics to benchmark, can we add (select with space)?

Agree with this suggestion. Added in the next commit.

Daanvdplas reviewed (commented):
Thanks for the last UX improvements!

> This goes against the benefit of the parameter menu that we defined in the spec. The benefit of the menu is to allow user to update the parameter again if they mistype. Reason for that because benchmarking has more parameters compared to other commands. Hence, this gives the user more control to the parameters of the command.

I disagree, one provides inputs to not have to provide them again. This also works for the other interactive UI's for the CLI.
